### DISCUSSION TITLE
QSV Support Revisited - SUCCESS!

### DISCUSSION THREAD

**Discussion by DZaremba14600** [automatic-ripping-machine/automatic-ripping-machine#1185] [General]
Over the past 10 months or more, I've documented my pursuit of using QSV in ARM.

I got a script to install ARM on Ubuntu 22.04 non-docker so that I could use QSV and I've touched on attempts to get QSV working reliably in an ARM Docker container.

A few things happened recently and I found that I needed to revisit all of this.  Then end result is that I have QSV fully running in the ARM Docker container, most everything involved has been updated to the most recent supported versions, and its all working well enough that I' going to be using it for a mass project and thought I'd share for anybody's benefit, including the maintainer's, if they see fit. (1337-server,microtechno9000, shitwolfymakes, muckngrind4).

I've never contributed officially to any project, I don't really know the particulars of how to go about doing it, but this seemed to fix an issue, so I'm just putting it out there.

So here we go.

I have a system with an i7-12700K so that has QSV support.

I recently read about Intel ARC discrete GPU cards and read how well they perform for transcoding.  The A380 was said to be great for its price as an encoder, so I bought a Sparkle unit for about $100 and thought I'd give it a try.  I was curious about the claims that it did AV1 encoding and it it very well and very fast.

I decided to use my previous work on installing ARM on Ubuntu 22.04 with no Docker and just see how that would go.  Well, I found that my previous work was now out of date - HandBrake was a new version with new dependencies so that script now had issues completing, the Intel Media SDK project was now closed to further development and that script now had issues, and so on.

I decided to install Ubuntu 22.04, go step by step, and bring everything I had previously done up to date.

1)  I installed Ubuntu 22.04 and did a full update to bring it up to date.  I then installed the latest ARC drivers following Intel's instructions.  So far so good.
2) Intel Media SDK is no longer support and has been replaced by OneVPL, so I install the runtime per Intel's instructions.  This is found on the install instructions for the ARC380 card on Intel's site and its called ' install Compute Runtime, Media, and Mesa packages'.  These contain all the encoders and associated support for transcoding.   Again, so far so good.  I also install VAInfo and run that and verify the system is seeing all the entry point required for these libraries to work and that all looks well.  
3) I built and installed MakeMKV from source - that script worked just fine.  I also bought a perpetual license since it works great and I'm going to keep using it.  I'm one to support the effort.
4) I build and installed the latest HandBrake from source - I had to redo that scrip because of new dependencies for the new 1.8.x version.  I did notice a potential issue with .265 10-bit QSV but I'm still researching that.  But I have scripts to install both 1.7.x an 1.8.x dependencies and with the matching version of Handbrake itself.

At this point all looks and works well, both HandBrake 1.8.2 GUI and CLI.  The issue with 265 10-bit QSV is that the GUI HandBrake reports that encoder as available and uses it where the CLI HandBrake does not list the 265 10-bit QSV encoder as availble BUT it will use it if its specified.  The activity logs bear that out.  Probably a bug in the 'list encoders' logic.  HandBrake 1.7.3 works and repost as expected.

Bottom line, Handbrake 1.7.x and 1.8.x work as expected.

I used MakeMKV to pull a movie MKV off a blueray for testing.  I tested encoding the RAW MKV to an AV1 MKV.

Um....  wow.  just wow.  super, super, fast.  

The movie looks fantastic on my computer screen.  I don't remember all the particular settings, I can do another benchmark but it was fast and the picture looked fantastic and the file size was pleasantly small.  However, I'm playing my stuff though a Roku and that doesn't support AV1 coming from Jellyfin.

I try  .265 QSV and again wow.  Super HQ 1080p preset is running at 270 to 280 fps in the Handbrake activity log.  

Ok, so now let me try getting ARM going again on the non-docker Ubuntu install doing a 'manual' install.  I used my previous script and that still worked fine.  I needed to tweak something minor, but I quickly got it working. 

Now I decided to turn my attention to QSV in the ARM Docker container.

To make this story shorter I ended up doing the following:

1)  I updated the ARM dependencies Dockerfile to use an Ubuntu 22.04 jammy image as a BASE, instead of a 20.04 focal image. 
2)  I updated the install_mkv_hb_deps.sh script to install the Intel UHD/ARC drivers that support QSV as well as installing the Compute Runtime, Media, and Mesa packages mentioned above. - *** this is the key for QSV support ***
3)  I updated the install_mkv_hb_deps.sh script to allow installing dependencies for either HandBrake 1.7.x OR 1.8.x.
4)  I updated the install_handbrake.sh script to allow installing HandBrake 1.7.3 or 1.8.2.
I build the arm_dependencies image and use a tag (I use my initials) so that the image is created local.
I updated the main ARM Dockerfile to use the tag I just built arm_dependencies with so that it pulls the local copy, not the Docker Hub copy/version.

I now fire up the container using the tag I used to build.  I have a working docker container of ARM that is QSV enabled.

All together, the 'changes' are actually quite minimal.

I tested this on my i7-12700K and also  on my Jellyfin machine which uses an i7-9700, has the same Sparkle A380 ARC GPU card.  I was able to rip and transcode without a hitch.  

I use Portainer to make use of Docker Compose to run the container and also to attach a shell to the ARM container to check everything, especially HandBrake.

I've done both BluRay and UHD, and HDR is preserved.  

This is so fast I'll be running higher quality settings and just balancing my quality against my space requirements.

I'll be running a few more tests this evening on a third machine, another i7-12700K with another Sparkle A380.  The end goal is that after this holiday weekend, I'm going to be re-doing my entire collection, for a number of reasons.

If anybody has any questions, please feel free. 

If anybody would like what's changed or how to go about enabling this, let me know.  All you have to be able to do is build a Docker image.

Maintainers, If this would be on any help or interest to you, let me know the best way to get this to you for your examination.  Everything is commented.

Thanks and hopefully someone finds this of interest or of help.

David.

**Comment by DZaremba14600** [comment]
A secondary comment - through all of this work, I also found an easy way to use custom presets created in HandBrake in docker ARM.

This functionality does not require and modifications.  I'll be running my tests this evening using custom presets - one for Bluray and and one for UltraHD 4K HDR.  I can document this too.  Very useful, very easy, very cool.

Again, let me know if there's an interest.

Thanks.

**Comment by abrugsch** [comment]
I am very interested in the steps you did to make this happen. I'm not massively familiar with building docker images from scratch so more detail on each of the steps would be great. or maybe this could be raised as a github issue to be pulled into the official image. I'd like to use presets that use QSV for h265 but `qsv: not available on this system` is output on `HandBrakeCLI --version` on the official image.

I'd very much like to get this working since as you've pointed out, the speed difference is phenomenal

**Reply by SrIzan10** [reply]
does your card support the intel media driver? mine is too old for it (ivy bridge), so I can't use that. If you guys have alternatives (like VAAPI) it would be awesome!
https://github.com/intel/media-driver?tab=readme-ov-file#supported-platforms

**Reply by abrugsch** [reply]
Mine does. it's a Kaby Lake (7th gen) Pentium G series and fully supports h265 10-bit encoding. If I use a standalone handbrake docker container then the options are there (and they work!) but for sure, Ivy bridge won't have it. 
I *THINK* Ivy bridge does have some hardware encoding but really you want a SkyLake or later chip.

**Reply by SrIzan10** [reply]
this is the dockerfile i found on a github issue on this exact repo.
make sure to build the dockerfile in a good enough chip, because it will compile some stuff.
```
FROM 1337server/automatic-ripping-machine:latest 
LABEL desc="ARM w/ Intel Quick Sync Video Support"
RUN apt update &&\
    apt install -y libmfx1 libmfx-tools libmfx-dev vorbis-tools wget &&\
    mkdir -p quicksync && cd quicksync &&\
    git clone https://github.com/intel/libva.git libva &&\
    cd libva &&\
    ./autogen.sh &&\
    make && make install &&\
    cd .. &&\
    wget https://github.com/Intel-Media-SDK/MediaSDK/releases/download/intel-mediasdk-21.3.5/MediaStack.tar.gz &&\
    tar -xvf MediaStack.tar.gz &&\
    cd MediaStack &&\
    bash install_media.sh &&\
    cd .. &&\
    git clone https://github.com/Intel-Media-SDK/MediaSDK msdk &&\
    cd msdk &&\
    mkdir build && cd build &&\
    cmake .. &&\
    make && make install &&\
    cd .. &&\
    rm -rf quicksync
```
then pass `/dev/dri/renderD128:/dev/dri/renderD128` and `/dev/dri/card0:/dev/dri/card0` as devices, just like the disc readers!
hope that you don't hit any roadblocks :wink:

**Reply by DZaremba14600** [reply]
Thanks for the responses!

Yes, it makes a huge difference in time and because I can set the RF/QC setting to a higher quality and it ends up working really well.

I'm in the process of re-processing my library in total.  I processed 23 blu-ray disks from ripping to completion in about 23 hours total.  And that included a break when I went to work.  

It's made a huge difference for me in workflow.


Let me do this - 

I have 2 systems I am running with 3 Asus blu-ray readers between them.  This evening I will post up on here and document my hardware (they are both Alder Lake i7-12400K - with the integrated GPU but I have a 3rd system with a Coffee Lake processor that I've tested on.

This evening when I get home I will document my hardware.  I will also post the changed files (only the files changed) with an explanation of what the changes are and why.  I will also include my Docker Compose files.  I use Portainer and use the Docker Compose functions.

I will also document how to use a custom Handbrake preset in ARM (I've found it useful) and I'll post a description of my ARM setup.

I'm reprocessing my entire library for a variety or reasons and I wanted to use QSV to save time and I also wanted a totally consistent set of settings.


Srlzan10 - That script looks great, wish I had seen that before, however, the Intel MediaSDK is now a closed project and will no longer be updated.  The git project page announcing that also mentions some security issues.  So going forward the MediaSDK will support fewer and fewer newer gen CPU/GPU's.  I ended up going with the OneVPL project which is replacing MediaSDK.

And yes, the /dev/dri is VERY important, none of this works without it.  Since I'm using both the integrated GPU on the CPU and also a discrete Intel ARC card, I pass in both, handbrake sees both and it selects which GPU it wants to use.  Pretty cool.



I've been hoping somebody would find this useful, I had a lot to share and decided to wait to post to gauge interest before posting it all up.  I'll post a lot later this evening.

Thanks.

**Reply by DZaremba14600** [reply]
[ARM-QSV-Docker.zip](https://github.com/user-attachments/files/17092506/ARM-QSV-Docker.zip)

I apologize I was unable to get his posted sooner.

Ok, I'm going to document my environment and what I've done to get QSV working in ARM in the Docker Container.  I hope this is helpful, and if there are any questions, feel free to ask.

I'll first document my hardware, then the host software, and then the Docker Container.
I've included an attached .ZIP file with the files I'll be referring to.


Hardware:

I've got 3 systems in use.

I've got a homebuilt system with an i7-12400k processor, UHD770 integrated graphics on an MIS motherboard, Sparkle Intel ARC A380 discrete GPU, ASUS BW-16D1HT SATA blu-ray
I've got a homebuilt system with an i7-12400k processor, UHD770 integrated graphics on a Gigabyte Motherboard, Sparkle Intel ARC A380 discrete GPU, ASUS BW-16D1HT SATA blu-ray
I've got an HP ProDesk 600 G5 with an i7-9700 processor, UHD630 integrated graphics, Sparkle Intel ARC A380 Genie discrete GPU, ASUS BW-16D1HT SATA Blu-ray in a USB 3.0 external case

the i7-12400K systems were put into existing cases as upgrades and are my main systems and dual-boot Win11 and Ubuntu 22.04
the ProDesk 600 is dual-boot Win11 and Ubuntu 22.04 and is used as a small home server running Jellyfin Docker and some other containers I play with

The ARC discrete GPUs were said to have very fast QSV encoding and supported QSV AV1 encoding.  I got one card and was impressed, so I got 2 more.  There were each about 120.00.  Cant complain.  I got them primarily to experiment with AV1 but also for the fast QSV support.  I have not been disappointed.

Software:

For the sake of ARM, I am running Ubuntu 22.04 LTS.  On this host I have run ARM non-Docker (and still can if I choose), but I've decided Docker is the way to go, hence this series of threads.  

Also on this host I run Docker, Portainer, Handbrake GUI (built from source, non-snap) and MakeMKV (build from source), Media Info, and ARM in a Docker Container.

I'm going to document how I set up my system, not to have you do the same, but for you to see what I did from start to finish.

I've included a .ZIP file that contains the following:

- an ARM folder - this folder is a folder structure which contain CHANGED files in the ARM project.  if the ARM project files were downloaded in a ZIP or with git into a folder named ARM, these files would replace like files in the folder structure.  All other files would remain the same.

- a 'system setup' folder which contains scripts I use to go from a clean Ubuntu 22.04 install that's been updated to a completed usable ARM system (except perhaps for personal tweaks or hardware differences).  But pretty close to ready.

- misc additional files to be explained.

The contents of this .ZIP file would go into the home folder.

Setting up the Host machine:

I install Ubuntu 22.04 LTS using Ventoy.  Just a straightforward install, defaults except for personal settings (hostname, timezone, user, password, etc).
I reboot and let 'Software Updater' update the system.  I've also used 'apt update, apt/upgrade).  I reboot when that's finished.
At this point I have a newly installed, fully updated 22.04 LTS system.

I would then basically get all the files in the ZIP folder 'system setup' into my home directory.

I then run the script 'sudo ./Install_Intel_ARC_GPU.sh'.  That installs an Intel GPU repository, and then installs all the packages for both GPU's in the system.  This script installs support for both the 12th (or 13th) gen processors integrated UHD GPU's AND the Intel ARC GPU.  Both are supported by what's installed.  These packages are the Compute Runtime, Media, and Mesa packages and also the development packages (just in case).  These packages include the intel-media-va-driver-non-free, the OneVPL runtimes, Media support, and Mesa drivers.  This is per Intel.  Also installed is the VAInfo utility (vainfo) and the Intel GPU Top utility (intel_gpu_top).  Running vainfo should display all the entry points needed for all the encoders/decoders supported.  It's quite the list.  I then reboot.

I install a package that give me a GUI for user and group management 'sudo apt-get install gnome-system-tools'.

The next two steps install MakeMKV GUI and Handbrake GUI.  I want those easily available so I can do custom Presets in Handbrake and also do easy testing.

I then run the script 'sudo ./Install_MakeMKV.sh'.  This installs all the dependencies for MakeMKV gui version, downloads the source and builds and installs the package.  It places an icon in my applications list and I add the icon to my favorites bar.

I then fire up MakeMKV and make sure all is ok.

I then run the script 'sudo ./Install-Handbrake-1-7.sh' OR 'sudo ./Install-Handbrake-1-8.sh'.  Not Both.  These scripts will install all the dependencies for the 1.7 or 1.8 branch of Handbrake and will then download the sourcea dn build and install the selected package.  It place an icon in my applications list and I add the icon to my favorites bar.

I then fire up Handbrake and make sure all is ok.  I should see all the encoders in the video encoder dropdown.  NOTE!! - This is the case for both 1.7 and 1.8, HOWEVER, 1.8 seems to have a bug such that running 'HandBrakeCLI --help | grep -A12 "Select video encoder"' displays the available encoders but will not display the 'qsv_h265_10bit' encoder in the list. All other QSV encoders show.  It shows in the 1.8 GUI, it shows in 1.7 GUI, it show in 1.7 CLI, but not in 1.8 CLI.  HOWEVER, if I use that encoder from the CLI in a CLI operation, the encoder is used.  In the .ZIP file, I included an ARM log for processing Planet of the Apes 1968 bluray, and the log shows that encoder selected and used.  It would seem that not showing in the CLI encoder list is a bug of some kind.

I then run the script 'sudo ./Install_Docker.sh'.  This installs Docker, and also installs Portainer.  I use Portainer for the easy access to a command shell in the containers.  This script installs an older version of Docker and it used to install an older version of Portainer to work around a bug, but I changed the script to use the latest working version or Portainer and didn't change the script to install the latest Docker.  Any system package update will update Docker accordingly.  I'll change the script when I get to it.

I then fire up a browser and go into Portainer, set up the Admin user and check everything.

At this point everything should be working and ready for ARM Docker container QSV support.

I need to pull down the ARM project source and I can do this in one of two ways - a ZIP file download from the Git project page, or I can use Git.  Note, If I use GIT, it pulls down the latest version.  This was 2.8. I had problems with 2.8 unrelated to QSV and didn't want to pursue that right now, so I downloaded version 2.7.2 as a zip and unpacked that into my home directory into a folder named 'arm'.  I also have to download the arm/arm-dependencies folder from the project page as a zip (its a sub-project) and place it into the arm folder.  I have a script I have provided that does into the home directory and I run 'sudo pull-arm.sh'.  This script uses git to pull down the project, the sub-project and readies that new 'arm' folder structure for my edits.

At this point I can configure the system for ARM Docker.  I can either do the manually, step by step, as documented, or I can tun the docker-setup.sh script.  I have found that the script works well for me.  So I get a copy of docker-setup.sh into my home directory, and flag it for execute.  It's important to remember that the UID that ARM uses in the container MUST be a member of the 'render' group for QSV to work.  That user, 'arm', needs 'render' group access in able to use the GPU to do any render functions.  If docker-setup.sh is run as is, don't forget to add the 'arm' user to the 'render' group.  I made that change to docker-setup.sh.  I included that modified script in the root of the zip file (what would be the home directory).  I modified line 49 and added the 'render' group to the usermod command.

I then run the script 'sudo ./docker-setup.sh'  This sets up the arh ARM Docker Container environment and all that's required.  I usually configure the 'arm' user to use a bash shell.

At this point I reboot and I log in as 'arm' user.  I check that everything looks ok, and that I can access ARM though a browser.  I don't know if it's me, or what, but I almost always have some issue with permissions on the home directory for 'arm' user that's related to the new folders.  I usually end up running 'sudo chown -R arm:arm arm' from the /home directory.  I sometimes have to start ARM with localhost:8080/setup and let that set permissions.  I hate to say it, but there's something inconsistent there.  But, get the owner set on arm to 'arm' and all folder and files included and you should be ok.

Once you have ARM docker running with the tag:latest, you can begin making the modifications to the ARM source.  This is very straight forward.  There just aren't many changes.  So, here we go.

First, I change the base used in arm/arm-dependencies/Dockerfile from an older 20.04 ubuntu image 'phusion/baseimage:focal-1.2.0'to a newer 22.04 ubuntu image 'phusion/baseimage:jammy-1.0.4'.  This is found on line 3 of arm/arm-dependancies/Dockerfile.  I've got it commented in what I provided.

Second, I change the tag on the base used in arm/Dockerfile from automaticrippingmachine/arm-dependencies:1.1.8 to automaticrippingmachine/arm-dependencies:qsv .  This is found on line 3 of arm/Dockerfile.  I've got it commented in what I provided.

Third, I add a large number of packages to script arm/arm-dependancies/scripts/install_mkv_hb_deps.sh . My additions are added after line 20.  There is an initial block for MakeMKV that needs no changes.  After than, around line 20, I added the same Intel GPU repository and media driver, OneVPL Compute Runtime, Media, and Mesa packages, and misc utils that were installed on the host at the beginning.  This will interface the GPU functionality in the container to the GPU out on the host.  I included VAInfo and intel_gpu_top so they can be run inside the container from a command line from an attached shell (as in Portainer).  These are just used to verify things are working, they're not required.
Beginning at line 64, I have logic to install HandBrake dependencies.  There's logic to install either 1.7.3 or 1.8.x dependencies, just comment in/out the version you want.  It's explained in the comments. The two branches have different sets of dependencies.

Forth, I modify arm/arm-dependancies/scripts/install_handbrake.sh . THe only change I make here is at line 28, I add login to set the handbrake version to something other than the latest release (1.8.2), for example if you want to install 1.7.3.  This version must match the version branch set in install_mkv_hb_deps.sh line 64.  They need to match.

And there you have it.  4 files changed, 2 Dockerfiles changed, very minor, and 2 scripts changed, not so minor.

All that's left is to build the images.  I build the arm/arm-dependencies as its own image first with tag ':qsv', and then build the main ARM image second with the same tag of ':qsv'.  These keep it separate from ':latest'.  I use the script 'build-arm-qsv.sh'.  I run it from the home directory 'sudo ./build-arm-qsv.sh'.  It takes a while to build, download a bunch of stuff, and ends up with a final image in Docker of about 2.9gig.

Well, that's it.  From a fresh install to a finished ARM docker container with working QSV support.

I'll post this up and then I'll follow it up with the docker-compose file for portainer and custom preset support and the handbrake command line I use in the handbrake config screen and what it does, just for fun.

This has been way long, but it should explain why I did what I did.

I really hope this helps someone and maybe even the maintainers.

Feel free with any questions or comments.

Sorry this was so long, but I wanted to be as complete as possible.

Thanks for your time.

DZaremba14600

**Reply by padenbeecher** [reply]
I cannot replicate this functionality for the life of me. I have been able to get your build off the ground, verify H264/265 and AV1 capabilities in shell, but it fails every time I get to hardware encoding on a rip. In my ARM log, I see it failing with:

[matroska @ 0x7fe250687440] Application provided invalid, non monotonically increasing dts to muxer in stream 0: 33 >= 0
ERROR: avformatMux: track 0, av_interleaved_write_frame failed with error 'Invalid argument'

My build is a TrueNAS Scale Docker Container with an A310 GPU, and a W-2123 if that changes anything. I found this thread that suggests trying a nightly build of Handbrake:

https://github.com/HandBrake/HandBrake/issues/6786

With no luck resolving it still. Do you have any ideas on what I am missing?
